### ISSUE TITLE
data race around YAML encoder/docs

### ISSUE THREAD

**Issue by smira** [talos-systems/talos#3380]
### Logs

```
==================
WARNING: DATA RACE
Write at 0x000004c16710 by goroutine 29:
  reflect.typedmemmove()
      /toolchain/go/src/runtime/mbarrier.go:177 +0x0
  reflect.Value.Set()
      /toolchain/go/src/reflect/value.go:1569 +0x13b
  github.com/talos-systems/talos/pkg/machinery/config/encoder.readExample()
      /src/pkg/machinery/config/encoder/documentation.go:248 +0x270
  github.com/talos-systems/talos/pkg/machinery/config/encoder.populateExamples()
      /src/pkg/machinery/config/encoder/documentation.go:262 +0x17b
  github.com/talos-systems/talos/pkg/machinery/config/encoder.renderExample()
      /src/pkg/machinery/config/encoder/documentation.go:182 +0x1e7
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:187 +0x102e
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:251 +0x31c
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.(*Encoder).Marshal()
      /src/pkg/machinery/config/encoder/encoder.go:29 +0x58
  github.com/talos-systems/talos/pkg/machinery/config/encoder.(*Encoder).Encode()
      /src/pkg/machinery/config/encoder/encoder.go:41 +0x55
  github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1.(*Config).Bytes()
      /src/pkg/machinery/config/types/v1alpha1/v1alpha1_provider.go:73 +0x5a
  github.com/talos-systems/talos/pkg/resources/config.(*MachineConfig).DeepCopy()
      /src/pkg/resources/config/machine_config.go:69 +0x73
  github.com/talos-systems/os-runtime/pkg/state/impl/inmem.(*ResourceCollection).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/inmem/collection.go:73 +0x3d8
  github.com/talos-systems/os-runtime/pkg/state/impl/inmem.(*State).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/inmem/inmem.go:43 +0x88
  github.com/talos-systems/os-runtime/pkg/state/impl/namespaced.(*State).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/namespaced/namespaced.go:47 +0xde
  github.com/talos-systems/os-runtime/pkg/state.(*coreWrapper).Get()
      <autogenerated>:1 +0xdb
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:172 +0x217
  github.com/talos-systems/talos/internal/app/machined/pkg/controllers/v1alpha1.(*TimeStatusController).Run()
      /src/internal/app/machined/pkg/controllers/v1alpha1/time_sync.go:65 +0x50d
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).runOnce()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:325 +0x196
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).run()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:284 +0x366
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*Runtime).Run.func1()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/runtime.go:115 +0xa4

Previous read at 0x000004c16711 by goroutine 170:
  [failed to restore the stack]
```

```
WARNING: DATA RACE
Read at 0x000004c16620 by goroutine 29:
  github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1.(*Bond).Doc()
      <autogenerated>:1 +0x4f
  github.com/talos-systems/talos/pkg/machinery/config/encoder.getDoc()
      /src/pkg/machinery/config/encoder/documentation.go:132 +0xad
  github.com/talos-systems/talos/pkg/machinery/config/encoder.populateExamples()
      /src/pkg/machinery/config/encoder/documentation.go:254 +0x64
  github.com/talos-systems/talos/pkg/machinery/config/encoder.renderExample()
      /src/pkg/machinery/config/encoder/documentation.go:182 +0x1e7
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:187 +0x102e
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:251 +0x31c
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.addToMap()
      /src/pkg/machinery/config/encoder/encoder.go:280 +0xa8
  github.com/talos-systems/talos/pkg/machinery/config/encoder.toYamlNode()
      /src/pkg/machinery/config/encoder/encoder.go:210 +0xfa4
  github.com/talos-systems/talos/pkg/machinery/config/encoder.(*Encoder).Marshal()
      /src/pkg/machinery/config/encoder/encoder.go:29 +0x58
  github.com/talos-systems/talos/pkg/machinery/config/encoder.(*Encoder).Encode()
      /src/pkg/machinery/config/encoder/encoder.go:41 +0x55
  github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1.(*Config).Bytes()
      /src/pkg/machinery/config/types/v1alpha1/v1alpha1_provider.go:73 +0x5a
  github.com/talos-systems/talos/pkg/resources/config.(*MachineConfig).DeepCopy()
      /src/pkg/resources/config/machine_config.go:69 +0x73
  github.com/talos-systems/os-runtime/pkg/state/impl/inmem.(*ResourceCollection).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/inmem/collection.go:73 +0x3d8
  github.com/talos-systems/os-runtime/pkg/state/impl/inmem.(*State).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/inmem/inmem.go:43 +0x88
  github.com/talos-systems/os-runtime/pkg/state/impl/namespaced.(*State).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/state/impl/namespaced/namespaced.go:47 +0xde
  github.com/talos-systems/os-runtime/pkg/state.(*coreWrapper).Get()
      <autogenerated>:1 +0xdb
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).Get()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:172 +0x217
  github.com/talos-systems/talos/internal/app/machined/pkg/controllers/v1alpha1.(*TimeStatusController).Run()
      /src/internal/app/machined/pkg/controllers/v1alpha1/time_sync.go:65 +0x50d
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).runOnce()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:325 +0x196
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*adapter).run()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/adapter.go:284 +0x366
  github.com/talos-systems/os-runtime/pkg/controller/runtime.(*Runtime).Run.func1()
      /.cache/mod/github.com/talos-systems/os-runtime@v0.0.0-20210315190223-7b3d14457439/pkg/controller/runtime/runtime.go:115 +0xa4

Previous write at 0x000004c16620 by goroutine 170:
  [failed to restore the stack]
```
### Environment

- Talos version: [`talosctl version --nodes <problematic nodes>`]
- Kubernetes version: [`kubectl version --short`]
- Platform:

**Comment by smira** [comment]
Even with incomplete trace, it seems evident that we have a problem with using shared value for Doc which is being written to by the encoder.

**Comment by smira** [comment]
@Unix4ever assigning this to you as you wrote most of that code

**Comment by Unix4ever** [comment]
> Even with incomplete trace, it seems evident that we have a problem with using shared value for Doc which is being written to by the encoder.

It's not written there, or at least shouldn't be written. We probably need to copy the value.

**Comment by smira** [comment]
Closed via #3381
